Tests/AsyncHTTPClientTests/HTTPClientTests.swift

```diff
@@ -2533,4 +2533,59 @@ class HTTPClientTests: XCTestCase {
         XCTAssertEqual(info.connectionNumber, 1)
         XCTAssertEqual(info.requestNumber, 1)
     }
+
+    func testBackpressure() {
+        class BackpressureResponseDelegate: HTTPClientResponseDelegate {
+            typealias Response = Void
+            var count = 0
+            var processingBodyPart = false
+            var didntWait = false
+            var lock = Lock()
+
+            init() {}
+
+            func didReceiveHead(task: HTTPClient.Task<Response>, _ head: HTTPResponseHead) -> EventLoopFuture<Void> {
+                return task.eventLoop.makeSucceededFuture(())
+            }
+
+            func didReceiveBodyPart(task: HTTPClient.Task<Response>, _ part: ByteBuffer) -> EventLoopFuture<Void> {
+                self.lock.withLock {
+                    // if processingBodyPart is true then previous body part is still being processed
+                    // XCTAssertEqual doesn't work here so store result to test later
+                    if processingBodyPart == true {
+                        didntWait = true
+                    }
+                    processingBodyPart = true
+                    count += 1
+                }
+                // wait one second before returning a successful future
+                return task.eventLoop.scheduleTask(in: .milliseconds(200)) {
+                    self.lock.withLock {
+                        self.processingBodyPart = false
+                        self.count -= 1
+                    }
+                }.futureResult
+            }
+
+            func didReceiveError(task: HTTPClient.Task<Response>, _ error: Error) {}
+            func didFinishRequest(task: HTTPClient.Task<Response>) throws {}
+        }
+
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 5)
+        let client = HTTPClient(eventLoopGroupProvider: .shared(elg))
+        defer {
+            XCTAssertNoThrow(try client.syncShutdown())
+            XCTAssertNoThrow(try elg.syncShutdownGracefully())
+        }
+
+        let data = Data(count: 65273)
```

Comment:
out of interest: Why are we uploading data here? Aren't we testing download streaming?

Reply:
Easiest way to get it to download a sizeable chunk of data was upload data to HTTPBin. HTTPBin just returns it back.

Reply:
If I replace this with an HTTPBin entry
```
            case "/chunks":
                let data = Data(count: 8000)
                var buf = context.channel.allocator.buffer(capacity: data.count)
                buf.writeBytes(data)
                context.write(wrapOutboundOut(.head(HTTPResponseHead(version: HTTPVersion(major: 1, minor: 1), status: .ok))), promise: nil)
                context.writeAndFlush(wrapOutboundOut(.body(.byteBuffer(buf))), promise: nil)
                context.writeAndFlush(wrapOutboundOut(.body(.byteBuffer(buf))), promise: nil)
                context.writeAndFlush(wrapOutboundOut(.body(.byteBuffer(buf))), promise: nil)
                context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
                return
```
would that work also to ensure that the number of body parts is greater than 1?

Reply:
I see, that makes sense, thank you. I would probably create a route that returns a fixed amount of bytes.

Maybe something like `/send-zeroes/1000000` would send 1 million zeroes. Then you can just use

```
context.channel.allocator.buffer(repeating: 0, count: count)
```

And btw, the faster and shorter way of writing

```swift
                let data = Data(count: 8000)
                var buf = context.channel.allocator.buffer(capacity: data.count)
                buf.writeBytes(data)
```

is

```swift
let buf = context.channel.allocator.buffer(repeating: 0, count: 8000)
```

Reply:
Set it 100000, because if the test is successful it will take quite a while if you set the download size too high.

Reply:
@adam-fowler Ok, 150,000 should be enough, barring AHC changes, the max is 64k.